compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ImportParameters.cpp

```diff
@@ -0,0 +1,419 @@
+using ParameterIndex = std::unique_ptr<iree_io_parameter_index_t,
+                                       void (*)(iree_io_parameter_index_t *)>;
+
+static FailureOr<FileHandle> openArchiveFile(ModuleOp moduleOp,
+                                             StringRef archivePath) {
+  iree_allocator_t hostAllocator = iree_allocator_system();
+
+  // Open the archive (hopefully mapped).
+  auto fileOrErr = llvm::MemoryBuffer::getFile(
+      archivePath, /*IsText=*/false, /*RequiresNullTerminator=*/false,
+      /*IsVolatile=*/false, /*Alignment=*/std::nullopt);
+  if (std::error_code error = fileOrErr.getError()) {
+    llvm::errs() << "cannot open archive input file '" + archivePath +
+                        "': " + error.message();
+    return failure();
+  }
+  auto file = std::move(fileOrErr.get());
+
+  // A callback issued when a file is released to destroy the file.
+  iree_io_file_handle_release_callback_t fileReleaseCallback;
+  fileReleaseCallback.fn =
+      +[](void *user_data, iree_io_file_handle_primitive_t handle_primitive) {
+        delete reinterpret_cast<llvm::MemoryBuffer *>(user_data);
+      };
+  fileReleaseCallback.user_data = file.get();
+
+  // Wrap the archive in a file handle.
+  iree_io_file_handle_t *fileHandle = nullptr;
+  if (failed(handleRuntimeError(
+          moduleOp,
+          iree_io_file_handle_wrap_host_allocation(
+              IREE_IO_FILE_ACCESS_READ,
+              iree_make_byte_span(const_cast<char *>(file->getBufferStart()),
+                                  file->getBufferSize()),
+              fileReleaseCallback, hostAllocator, &fileHandle),
+          "unable to wrap archive memory buffer"))) {
+    return failure();
+  }
+  file.release(); // now owned by the fileHandle
+
+  return FileHandle(fileHandle, iree_io_file_handle_release);
+}
+
+static LogicalResult
+loadParameterIndex(ModuleOp moduleOp, StringRef path,
+                   iree_io_parameter_index_t *parameterIndex) {
+  // Open the archive file (hopefully mapping it).
+  auto fileHandle = openArchiveFile(moduleOp, path);
+  if (failed(fileHandle))
+    return failure();
+
+  // Parse the archive as a particular format.
+  // TODO(benvanik): centralize this type selection logic in iree/io/.
+  StringRef format = llvm::sys::path::extension(path);
+  if (format == ".gguf") {
+    return handleRuntimeError(
+        moduleOp, iree_io_parse_gguf_index(fileHandle->get(), parameterIndex),
+        "parsing gguf file");
+  } else if (format == ".irpa") {
+    return handleRuntimeError(
+        moduleOp, iree_io_parse_irpa_index(fileHandle->get(), parameterIndex),
+        "parsing irpa file");
+  } else if (format == ".safetensors") {
+    return handleRuntimeError(
+        moduleOp,
+        iree_io_parse_safetensors_index(fileHandle->get(), parameterIndex),
+        "parsing safetensors file");
+  } else {
+    llvm::errs() << "unsupported archive file format: " << path << "\n";
```

Comment:
nit: wrap strings in quotes when logging them, so empty strings stand out
```suggestion
    llvm::errs() << "unsupported archive file format: '" << path << "'\n";
```